### PR TITLE
Complete API rewrite

### PR DESCRIPTION
**Pull request by tusbar**
TODO:

- [x] Allow passing `etag` and `lastCheckedAt`
- [x] Extract etag from headers
- [x] Extract links for "index of" pages
- [x] Use `base` header when available in "index of" pages
- [x] Extract filenames encoded with rfc2047 from content-disposition header  
- [x] Analyze { inputType: 'path' } tokens
- [x] Allow analyzing files and directories directly from the command line
- [x] Flatten tree and match file patterns
- [x] Add (some) tests
- [x] Add atom feeds
- [x] Add checksum for all matched files (path + url)
- [x] Revert what’s been read for unmatched index-of and atom types
- [x] Stop overriding the type on error: the error property is enough
- [x] Allow requesting urls with special characters in them
- [x] Test archives (requires https://github.com/travis-ci/apt-package-whitelist/pull/4235 to be merged)
- [x] Add code coverage
- [x] Allow reverting archives so we can treat them as regular files any maybe switch the file type if the extract failed
- [x] Treat large files as regular files with a flag saying they were not downloaded, not as errors, maybe just warnings
- [x] Rewrite readme
- [x] Add connection and download timeouts
- [x] Add option for max download size
---

Known issues:

- [x] ~An html file’s body information will not be accurate as it has already been read by the html parser.~ 
- [x] ~Inferring the type from the first chunk is broken when the first chunk is too small. `file-types` requires up to 4100 bytes to infer the type properly, when we have less, this is broken (https://geobretagne.fr/pub/dreal_b/habitat/RPLS%202016%20Principaux%20indicateurs.xlsx is treated as a zip file)~

---

Non-working links:

- [x] ~http://sig.pays-de-dinan.org/pays/download.php?param=zone_plu~
- [x] ~https://geobretagne.fr/pub/dreal_b/habitat/RPLS%202016%20Principaux%20indicateurs.xlsx~

### CODE REVIEW

**Comment by jdesboeufs** [comment]
You can also consider supporting `Last-Modified` and its `If-Modified-Since` header counterpart.

**Comment by tusbar** [comment]
@jdesboeufs https://github.com/inspireteam/plunger/pull/7/files#diff-4033a7e50d6c3ddd177f61fec97c7f48R11

File: lib/util/request.js [hunk 1]
```diff
@@ -27,8 +27,7 @@ function request(location, headers, _redirections = []) {
       host,
       path: encodeURI(path),
       headers: Object.assign({}, headers, {
-        'Accept-Encoding': 'gzip',
-        'User-Agent': 'plunger/1.0'
+        'Accept-Encoding': 'gzip'
```
Comment (jdesboeufs) [inline lib/util/request.js]: Trailing comma? 😹

Comment (tusbar) [reply lib/util/request.js]: I’d rather die 💀

**Comment by jdesboeufs** [comment]
A resource may have been updated between two errored checks.
But you should talk about last *successful* check.

**Comment by tusbar** [comment]
Of course!

**Comment by tusbar** [comment]
@jdesboeufs I think we’re good for a first try, what do you think?

**Pull request merged**